Unable to run the same (failing) test twice

##### Issue Overview

When executing the same test twice, without cleaning `target` folder between the executions (e.g. from IDEA), Test Bed will fail when trying to store project under test within the project folder.

##### Expected Behaviour

Test should be executed successfully (and a folder of the project under test persisted without error).

##### Current Behaviour

Following stacktrace occurs when executing test for the second time from IDEA
<details>
 <summary>Stacktrace from IDEA</summary>

```
java.lang.IllegalStateException: Maven build has failed, see logs for details

	at org.arquillian.smart.testing.ftest.testbed.project.ProjectBuilder.build(ProjectBuilder.java:62)
	at org.arquillian.smart.testing.ftest.testbed.project.ProjectBuilder.run(ProjectBuilder.java:96)
	at org.arquillian.smart.testing.ftest.testbed.project.ProjectBuilder.run(ProjectBuilder.java:92)
	at org.arquillian.smart.testing.ftest.affected.LocalChangesAffectedTestsSelectionExecutionFunctionalTest.should_only_execute_tests_related_to_multiple_local_changes_in_production_code_when_affected_is_enabled(LocalChangesAffectedTestsSelectionExecutionFunctionalTest.java:59)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed$1.evaluate(TestBed.java:85)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
java.lang.RuntimeException: java.nio.file.FileAlreadyExistsException: target/projects/smart-testing-dogfood-repo_LocalChangesAffectedTestsSelectionExecutionFunctionalTest_should_only_execute_tests_related_to_multiple_local_changes_in_production_code_when_affected_is_enabled

	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed.copyTmpProjectInTarget(TestBed.java:70)
	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed.failed(TestBed.java:49)
	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed.failedQuietly(TestBed.java:144)
	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed.access$200(TestBed.java:20)
	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed$1.evaluate(TestBed.java:89)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.nio.file.FileAlreadyExistsException: target/projects/smart-testing-dogfood-repo_LocalChangesAffectedTestsSelectionExecutionFunctionalTest_should_only_execute_tests_related_to_multiple_local_changes_in_production_code_when_affected_is_enabled
	at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:551)
	at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:253)
	at java.nio.file.Files.copy(Files.java:1274)
	at org.arquillian.smart.testing.ftest.testbed.rules.TestBed.copyTmpProjectInTarget(TestBed.java:67)
	... 21 more
```
</details>

##### Steps To Reproduce

Run selected test twice in IDEA. If it's failing nothing else is needed - it will fail with different error - the same as in detailed stacktrace above. Alternatively, for sucessful test, set `test.bed.project.persist` system property to ` true` prior to running the test twice in a row.

##### Additional Information

We can fix it in at least two ways:

- clean the project if already exists
- use timestamp as part of the directory structure

with latter, we persist all the executions before wiping `target` directory. It's a similar approach to how Arquillian persists test deployments so might be worth a shot.

--- a/docs/developing.adoc
+++ b/docs/developing.adoc
@@ -107,7 +107,7 @@ for example:
 You can execute embedded build in Test Bed in the debug mode.
 This gives better analysis of the build execution, as it lets us connect to the build execution to analyze behaviour of our extension in the runtime by debugging both Maven execution as well as Surefire Provider.
 
-===== Features
+===== Built-in DSL features
 
 You can use project configuration DSL for enabling debug mode directly in the test code:
 
@@ -134,15 +134,15 @@ Following system properties are available:
 If both system property and the programmatic option is used system property takes precedence.
 ====
 
-===== Project Persistent
-Project Persistence feature is included in Test Bed to store repository copy used by each and every test method to
-facilitate analysis in case of test failure.
+===== Storing project under test
+Project Persistence feature is included in Test Bed to store repository used test's embedded build. This lets you execute
+the same build outside of test execution for further analysis in case of failure.
 
 By default this feature is enabled to copy repository only in case of test failure.
 
 In order to copy repository for all tests irrespective of test result, explicitly set system property `test.bed.project.persist` to `true`.
 
-Persisted project are located in `target/projects` with name followed by naming convention:
+Persisted project are located in `target/test-bed-executions/[current timestamp]/` with name followed by naming convention:
 
 `getClass().getSimpleName() + "_" + name.getMethodName()`
 

--- a/functional-tests/test-bed/src/main/java/org/arquillian/smart/testing/ftest/testbed/rules/TestBed.java
+++ b/functional-tests/test-bed/src/main/java/org/arquillian/smart/testing/ftest/testbed/rules/TestBed.java
@@ -5,6 +5,7 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.time.Instant;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.logging.Logger;
@@ -50,10 +51,10 @@ public class TestBed implements TestRule {
     }
 
     private void copyTmpProjectInTarget() {
-        String path = "target" + File.separator + "projects";
+        String path = "target" + File.separator + "test-bed-executions" + File.separator + Instant.now().toEpochMilli();
         final File projectDir = new File(path);
         if (!projectDir.exists()) {
-            projectDir.mkdir();
+            projectDir.mkdirs();
         }
         if (targetRepoPerTestFolder != null) {
             final Path source = Paths.get(targetRepoPerTestFolder);

--- a/functional-tests/test-bed/src/main/java/org/arquillian/smart/testing/ftest/testbed/rules/TestBed.java
+++ b/functional-tests/test-bed/src/main/java/org/arquillian/smart/testing/ftest/testbed/rules/TestBed.java
@@ -69,7 +69,7 @@ public class TestBed implements TestRule {
             } catch (IOException e1) {
                 throw new RuntimeException(e1);
             }
-            LOGGER.info("copied test repository for failed test in: " + target);
+            LOGGER.info("Copied test repository to: " + target);
         }
     }
 

--- a/functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/testbed/ProjectPersistTest.java
+++ b/functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/testbed/ProjectPersistTest.java
@@ -1,61 +1,49 @@
 package org.arquillian.smart.testing.ftest.testbed;
 
 import java.io.File;
-import net.jcip.annotations.NotThreadSafe;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
 import org.arquillian.smart.testing.ftest.testbed.rules.GitClone;
 import org.arquillian.smart.testing.ftest.testbed.rules.TestBed;
 import org.junit.Assert;
 import org.junit.ClassRule;
 import org.junit.Rule;
 import org.junit.Test;
-import org.junit.contrib.java.lang.system.RestoreSystemProperties;
 import org.junit.runner.JUnitCore;
 import org.junit.runner.Result;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-@NotThreadSafe
 public class ProjectPersistTest {
 
-    @Rule
-    public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
-
     @Test
-    public void should_copy_failing_test_directory_in_target_if_test_is_failing() {
-        final Result result = JUnitCore.runClasses(ProjectPersistFail.class);
-
-        assertThat(result.wasSuccessful()).isFalse();
-        assertThat(new File("target/projects/smart-testing-dogfood-repo_ProjectPersistFail_should_fail")).isDirectory().exists();
-    }
-
-    @Test
-    public void temp_projects_should_copied_in_target_if_test_is_failing() {
+    public void should_store_project_under_test_to_two_separated_directories_when_the_same_test_fails_twice_in_a_row() throws IOException {
         final Result firstRun = JUnitCore.runClasses(ProjectPersistFail.class);
         final Result secondRun = JUnitCore.runClasses(ProjectPersistFail.class);
 
         assertThat(firstRun.wasSuccessful()).isFalse();
         assertThat(secondRun.wasSuccessful()).isFalse();
-        // it's rather shallow check, but if it's not equal it means second execution of the test failed for different reason
+        // it's rather shallow check, but if it's not equal it means second execution of the test failed for different reasons
         assertThat(secondRun.getFailures()).hasSameSizeAs(firstRun.getFailures());
-        assertThat(new File("target/projects/smart-testing-dogfood-repo_ProjectPersistFail_should_fail")).isDirectory().exists();
+
+        assertThat(findPersistedProjects("smart-testing-dogfood-repo_ProjectPersistFail_should_fail")).hasSize(2);
     }
 
     @Test
-    public void temp_test_projects_should_not_copied_in_target_if_test_is_passing() {
+    public void should_not_store_project_under_test_directory_when_test_is_passing() throws IOException {
         final Result result = JUnitCore.runClasses(ProjectPersistPass.class);
 
         assertThat(result.wasSuccessful()).isTrue();
-        assertThat(new File("target/projects/smart-testing-dogfood-repo_ProjectPersistPass_should_pass")).doesNotExist();
+        assertThat(findPersistedProjects("smart-testing-dogfood-repo_ProjectPersistPass_should_pass")).isEmpty();
     }
 
-    @Test
-    public void temp_test_projects_should_copied_in_target_if_test_is_passing_and_system_property_is_set() {
-        System.setProperty("test.bed.project.persist", "true");
-
-        final Result result = JUnitCore.runClasses(ProjectPersistPass1.class);
-
-        assertThat(result.wasSuccessful()).isTrue();
-        assertThat(new File("target/projects/smart-testing-dogfood-repo_ProjectPersistPass1_should_pass")).isDirectory();
+    private List<Path> findPersistedProjects(String projectName) throws IOException {
+        return Files.walk(new File("target" + File.separator + "test-bed-executions").toPath(), 2)
+            .filter(dir -> dir.toFile().getAbsolutePath().endsWith(projectName))
+            .collect(Collectors.toList());
     }
 
     public static class ProjectPersistFail {
@@ -85,16 +73,4 @@ public class ProjectPersistTest {
         }
     }
 
-    public static class ProjectPersistPass1 {
-        @ClassRule
-        public static final GitClone GIT_CLONE = new GitClone();
-
-        @Rule
-        public TestBed testBed = new TestBed(GIT_CLONE);
-
-        @Test
-        public void should_pass() throws Exception {
-            Assert.assertTrue(true);
-        }
-    }
 }

--- a/functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/testbed/ProjectPersistTest.java
+++ b/functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/testbed/ProjectPersistTest.java
@@ -21,10 +21,22 @@ public class ProjectPersistTest {
     public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
 
     @Test
-    public void temp_projects_should_copied_in_target_if_test_is_failing() {
+    public void should_copy_failing_test_directory_in_target_if_test_is_failing() {
         final Result result = JUnitCore.runClasses(ProjectPersistFail.class);
 
         assertThat(result.wasSuccessful()).isFalse();
+        assertThat(new File("target/projects/smart-testing-dogfood-repo_ProjectPersistFail_should_fail")).isDirectory().exists();
+    }
+
+    @Test
+    public void temp_projects_should_copied_in_target_if_test_is_failing() {
+        final Result firstRun = JUnitCore.runClasses(ProjectPersistFail.class);
+        final Result secondRun = JUnitCore.runClasses(ProjectPersistFail.class);
+
+        assertThat(firstRun.wasSuccessful()).isFalse();
+        assertThat(secondRun.wasSuccessful()).isFalse();
+        // it's rather shallow check, but if it's not equal it means second execution of the test failed for different reason
+        assertThat(secondRun.getFailures()).hasSameSizeAs(firstRun.getFailures());
         assertThat(new File("target/projects/smart-testing-dogfood-repo_ProjectPersistFail_should_fail")).isDirectory().exists();
     }
 

--- a/functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/testbed/ProjectPersistUsingPropertyTest.java
+++ b/functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/testbed/ProjectPersistUsingPropertyTest.java
@@ -1,0 +1,56 @@
+package org.arquillian.smart.testing.ftest.testbed;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+import net.jcip.annotations.NotThreadSafe;
+import org.arquillian.smart.testing.ftest.testbed.rules.GitClone;
+import org.arquillian.smart.testing.ftest.testbed.rules.TestBed;
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.RestoreSystemProperties;
+import org.junit.runner.JUnitCore;
+import org.junit.runner.Result;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@NotThreadSafe
+public class ProjectPersistUsingPropertyTest {
+
+    @Rule
+    public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
+
+    @Test
+    public void should_store_project_under_test_directory_when_test_is_passing_but_property_is_set() throws IOException {
+        System.setProperty("test.bed.project.persist", "true");
+
+        final Result result = JUnitCore.runClasses(ProjectPersistAnotherPass.class);
+
+        assertThat(result.wasSuccessful()).isTrue();
+        assertThat(findPersistedProjects("smart-testing-dogfood-repo_ProjectPersistAnotherPass_should_pass")).hasSize(1);
+    }
+
+    private List<Path> findPersistedProjects(String projectName) throws IOException {
+        return Files.walk(new File("target" + File.separator + "test-bed-executions").toPath(), 2)
+            .filter(dir -> dir.toFile().getAbsolutePath().endsWith(projectName))
+            .collect(Collectors.toList());
+    }
+
+    public static class ProjectPersistAnotherPass {
+        @ClassRule
+        public static final GitClone GIT_CLONE = new GitClone();
+
+        @Rule
+        public TestBed testBed = new TestBed(GIT_CLONE);
+
+        @Test
+        public void should_pass() throws Exception {
+            Assert.assertTrue(true);
+        }
+    }
+}

